File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)

        prefix_maxes = []
        prefix_max = 0

        suffix_maxes = []
        suffix_max = 0

        for i in range(n):
            prefix_maxes.append(prefix_max)
            suffix_maxes = [suffix_max] + suffix_maxes

            prefix_max = max(prefix_max, height[i])
            suffix_max = max(suffix_max, height[n-i-1])

        total = 0

        print(prefix_maxes)
        print(suffix_maxes)
        maxes = []

        for i in range(n):
            total += max(min(prefix_maxes[i], suffix_maxes[i]) - height[i], 0)
            maxes.append(min(prefix_maxes[i], suffix_maxes[i]) - height[i])

        print(maxes)

        
        return total

        
        


        

File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        vector<int>::iterator p1 = heights.begin();
        vector<int>::iterator p2 = heights.end() - 1;

        int max = 0;
        int curr;

        while(p1 < p2) {
            curr = (p2-p1) * std::min(*p1, *p2);
            std::cout << *p1 << " " << *p2 << " " << curr <<"\n";
            if(*p1<*p2) ++p1;
            else --p2;

            max = std::max(max, curr);
        } 

        return max;

    }
};
